README.md
Go Coordinate Daemon
====================

[![CircleCI](https://circleci.com/gh/diffeo/go-coordinate.svg?style=svg)](https://circleci.com/gh/diffeo/go-coordinate)
[![Docker Hub Repository](https://img.shields.io/docker/pulls/diffeo/coordinated.svg "Docker Hub Repository")](https://hub.docker.com/r/diffeo/coordinated/)

This package provides a reimplementation of the Diffeo Coordinate
(https://github.com/diffeo/coordinate) daemon.  It is fully compatible
with existing Python Coordinate code, and provides a REST interface
for Go and other languages.

* [Documentation index](doc/index.md)
* [Change history](doc/changes.md)

Overview
--------

Coordinate is a job queue system.  It is designed for repetitive tasks
with large numbers of inputs, where the inputs and outputs will be
stored externally and do not need to be passed directly through the
system, and where no particular action needs to be taken when a job
finishes.

Coordinate-based applications can define _work specs_, JSON or YAML
dictionary objects that define specific work to do.  A typical work
spec would name a specific Python function to call with a YAML
configuration dictionary, and the Python Coordinate package contains a
worker process that can run these work specs.  Each work spec has a
list of _work units_, individual tasks to perform, where each work
unit has a name or key and an additional data dictionary.  In typical
operation a work unit key is a filename or database key and the data
is used only to record outputs.  Read more about the
[data model](doc/model.md).

The general expectation is that there will be, at most, dozens of work
specs, but each work spec could have millions of work units.  It is
definitely expected that many worker processes will connect to a
single Coordinate daemon, and past data loads have involved 800 or
more workers talking to one server.

Installation
------------

From source:

    go get github.com/diffeo/go-coordinate/cmd/coordinated

Usage
-----

Run the `coordinated` binary.  With default options, it will use
in-memory storage and start a network server listening on ports 5932
and 5980.  Port 5980 provides the REST interface.

Go code should use the `backend` package to provide a command-line
flag to get a backend object, which will implement the generic
interface in the `coordinate` package.  Test code can directly create
a `memory` backend.  Most applications will expect to use the
`restclient` backend to communicate with a central Coordinate daemon.

5932 is the default TCP port for the Python Coordinate daemon, and
application configurations that normally point at that daemon should
work against this one as well.  Read more about
[Python compatibility](doc/python.md).

```sh
pip install coordinate
go get github.com/diffeo/go-coordinate/cmd/coordinated
$GOPATH/bin/coordinated &
cat >config.yaml <<EOF
coordinate:
  addresses: ['localhost:5932']
EOF
coordinate -c config.yaml summary
```

Docker
------

A [Coordinate daemon server image](https://hub.docker.com/r/diffeo/coordinated/)
is on Docker Hub:

```sh
docker run -p 5932:5932 -p 5980:5980 diffeo/coordinated
```

This is a single-binary image that only runs the daemon.  If you need
any additional command-line options, such as a persistent backend,
specify them directly after the image name.

```sh
docker run -d -p 5432:5432 postgres:9.5
docker run -d -p 5932:5932 -p 5980:5980 diffeo/coordinated \
    -backend postgres://172.17.0.1 -log-requests
```

The current CI setup has the Docker `latest` tag pointing at a
`master` commit from this repository.  You may want to specify a
specific version tag.  The earliest version tag in Docker Hub is
`diffeo/coordinated:0.4.2`.

Packages
--------

`cmd/coordinated` is the main process, providing the network service.
`jobserver` provides the RPC calls compatible with the Python
Coordinate system.  `cborrpc` provides the underlying wire transport.

`coordinate` describes an abstract API to Coordinate.  This API is
slightly different from the Python Coordinate API; in particular, an
Attempt object records a single worker working on a single work unit,
allowing the history of workers and individual work units to be
tracked.  `memory` is the in-memory implementation of this API,
`postgres` uses PostgreSQL, and `restclient` talks to a REST server.
`backend` provides a command-line option to choose a backend.

Future
------

The `Namespace.Workers()` call simply iterates all known workers, but
the implementation of the Python Coordinate worker will generate an
extremely large number of these.  This call is subject to unspecified
future change.

Go Coordinate version 0.3.0 adds a generic WorkUnitMeta structure,
which defines both the work unit priority and the earliest time it can
execute.  This replaces the priority field in a couple of contexts.  A
future version of Go Coordinate not before 0.4.0 will delete
`WorkUnit.Priority()` and `WorkUnit.SetPriority()`.


doc/changes.md
Major Changes
=============

0.4.13 (15 Sep 2018)
-------------------

* Added HTTP Request logging middleware.
  [#22](https://github.com/diffeo/go-coordinate/pull/22)
  [#23](https://github.com/diffeo/go-coordinate/pull/23)

0.4.12 (7 Sep 2018)
-------------------

* Added `.NumAttempts()` method as a faster way to do `len(w.Attempts())`
  [#21](https://github.com/diffeo/go-coordinate/pull/21)

0.4.11 (24 Jan 2018)
-------------------

* Added String method to WorkUnitStatus.
  ([#19](https://github.com/diffeo/go-coordinate/pull/19))

0.4.10 (2 Jan 2018)
-------------------

* Fix CircleCI build.
  ([#18](https://github.com/diffeo/go-coordinate/pull/18))

0.4.9 (2 Jan 2018)
------------------

* Switch to CircleCI 2.0.
  ([#17](https://github.com/diffeo/go-coordinate/pull/17))
* Switch to dep dependency management tool.
  ([#17](https://github.com/diffeo/go-coordinate/pull/17))

0.4.8 (21 Dec 2017)
-------------------

* Fix a data race.
  ([#16](https://github.com/diffeo/go-coordinate/pull/16))

0.4.7 (29 Aug 2017)
-------------------

* Switch docker image repository to Docker Hub.
  ([#15](https://github.com/diffeo/go-coordinate/pull/15))

0.4.6 (7 July 2017)
-------------------

* Bug fix for panic on nil logger.
  ([#14](https://github.com/diffeo/go-coordinate/pull/14))

0.4.5 (2 Jun 2017)
------------------

* Performance fix for the PostgreSQL backend when `max_retries` is used.
  ([#12](https://github.com/diffeo/go-coordinate/pulls/12))
* If `max_retries` causes some work units to get preemptively failed,
  `RequestAttempts` will return fewer than the `max_getwork` number of
  attempts, rather than attempting to get more work, on all backends.

0.4.4 (30 May 2017)
-------------------

* In the published Prometheus metrics, don't double-double-quote the
  "status" value.
  ([#11](https://github.com/diffeo/go-coordinate/pulls/11))

0.4.3 (30 May 2017)
-------------------

* Add a `max_retries` work spec parameter to limit work unit retries
  ([#8](https://github.com/diffeo/go-coordinate/pulls/8))
* Port the generic test suite to use
  `github.com/stretchr/testify/suite`, rather than the home-grown
  `cptest` program in this package.
  ([#9](https://github.com/diffeo/go-coordinate/pulls/9))
* Publish `coordinate summary` style metrics
  to [Promotheus](https://prometheus.io/).
  ([#10](https://github.com/diffeo/go-coordinate/pulls/10))

0.4.2 (4 May 2017)
------------------

* Requires Go 1.7 or later, and its standard `context` package
  ([#4](https://github.com/diffeo/go-coordinate/pulls/4))
* Add CircleCI integration, Quay reference Docker image
  ([#1](https://github.com/diffeo/go-coordinate/pulls/1),
  [#2](https://github.com/diffeo/pulls/2))
* PostgreSQL backend performance improvement for long-running systems
  with many old workers
  ([#3](https://github.com/diffeo/go-coordinate/pulls/3))
* The Go worker framework can take a list of runtime strings as
  parameters, it is not limited to `"go"`
  ([#5](https://github.com/diffeo/go-coordinate/pulls/5))

0.4.1 (15 Aug 2016)
-------------------

* Bug fix in Go decoding of CBOR-RPC responses.
* Bug fix adding work units in the PostgreSQL backend,
  where adding a work unit that already existed would affect all work
  specs with performance implications

0.4.0 (29 Jun 2016)
-------------------

* Bug fix relating to multiple concurrent Go workers and concurrent
  map access
* Bug fixes for continuous work units on the PostgreSQL backend,
  including fixes for intervals longer than 3600 seconds, intervals
  longer than 86400 seconds, and concurrent attempts to execute
  continuous work units
* Don't ignore a requested work spec name from Python
  `coordinate run_one --from-work-spec`

The `github.com/satori/go.uuid` library made its UUID parsing stricter
in this timeframe as well, so UUID values passed in work unit data or
elsewhere may be rejected if key bits do not conform to RFC 4122.

0.3.0 (8 Apr 2016)
------------------

* Adds the ability for work units to declare an earliest time they can
  run.  Changes `WorkSpec.AddWorkUnit()` to take a new `WorkUnitMeta`
  structure instead of the existing `priority` parameter, and adds a
  `time.Duration` parameter to `Attempt.Retry()`.  Generated `output`
  work units in chained work units can also declare a `delay` before
  running the next unit.  Makes corresponding (incompatible) changes
  to the REST API.
* Adds a worker framework so that Coordinate work specs can run task
  implementations written in Go.
* Add a `coordbench` tool to measure the performance of the Coordinate
  system.
* Several performance improvements for the PostgreSQL backend, aimed
  at the "millions of work units" scale.
* Switch tests from `gocheck` to `testify`.

0.2.0 (4 Jan 2016)
------------------

* Adds a REST server interface: start `coordinated -http :5980`.
* Adds a REST client backend: start `coordinated -backend
  http://localhost:5980/`.
* Adds a `runtime` key to work specs to allow workers for multiple
  language runtimes to coexist.
* Turns off detailed logging of every CBOR-RPC request and response.

0.1.2 (1 Dec 2015)
------------------

* Build changes to support getting a working static Docker container out.

0.1.1 (20 Nov 2015)
-------------------

* Work around a code change in the
  [codec](https://github.com/ugorji/go/codec) library.

0.1.0 (18 Nov 2015)
-------------------

* Initial release.


worker/worker.go
// Copyright 2016 Diffeo, Inc.
// This software is released under an MIT/X11 open source license.

// Package worker provides a library framework for processes that
// execute Coordinate work units.
package worker

import (
	"context"
	"fmt"
	"net"
	"os"
	"runtime"
	"sync"
	"time"

	"github.com/benbjohnson/clock"
	"github.com/diffeo/go-coordinate/coordinate"
	"github.com/satori/go.uuid"
)

// Worker is foo.
type Worker struct {
	// Namespace identifies the Coordinate namespace from which
	// work is performed.  This field is required when creating
	// a Worker.
	Namespace coordinate.Namespace

	// Tasks defines the tasks this Worker is capable of running.
	// Work specs must declare a runtime consistent with this
	// worker's runtime settings (by default, that's "runtime: go"),
	// and also have a "task:" field that names one of the tasks in
	// this map. If a work spec has no "task:", the work spec name
	// is looked up here instead.
	//
	// The task function is called with a context and a slice of
	// at least one attempt.  The context will be canceled when
	// the worker is stopped or if one of the attempts is nearing
	// is expiration time.  The worker can take any reasonable action
	// in response to being signaled, but generally it should stop
	// doing further work and mark all of the attempts as failed.
	//
	// There is guaranteed to be at least one attempt.  All attempts
	// are for the same worker and for the same work spec.
	Tasks map[string]func(context.Context, []coordinate.Attempt)

	// WorkerID provides the name of the worker as seen through the
	// Coordinate API.  If unset, a worker ID will be generated.
	WorkerID string

	// Concurrency states how many sets of attempts should run in
	// parallel.  If unset, uses runtime.NumCPU().
	Concurrency int

	// PollInterval states how often the worker should try to get
	// more work if the previous attempt(s) returned nothing.  If
	// unset, defaults to 1 second.
	PollInterval time.Duration

	// HeartbeatInterval states how often the worker should report
	// its status in the Coordinate worker data, and check for
	// work units that are about to expire.  If unset, defaults to
	// 15 seconds.
	HeartbeatInterval time.Duration

	// MaxAttempts limits the number of attempts that will be
	// returned; it is exactly the
	// coordinate.AttemptRequest.NumberOfWorkUnits parameter.  If
	// unset, use 100.  Note that individual work specs can also
	// limit the number of attempts by setting a "max_getwork"
	// key.
	MaxAttempts int

	// ErrorHandler is called when an error occurs in the worker
	// main loop.
	ErrorHandler func(error)

	// Clock defines a time source for the worker.  If the
	// Coordinate backend was created with an alternate time
	// source, this should match that time source.  Only test code
	// should need to set this.  If unset, uses a time source
	// backed by real wall-clock time.
	Clock clock.Clock

	// Runtimes sets the language runtimes that should be used
	// with this worker. When nil, a default value of ["go"] is
	// used.
	Runtimes []string

	// parentWorker is a saved Coordinate worker object with ID
	// WorkerID.
	parentWorker coordinate.Worker

	// childWorkers identifies the child workers.
	childWorkers map[string]coordinate.Worker

	// cancellations maps child worker ID to a cancellation function
	// for that worker's context.  These functions are specified to
	// be idempotent.
	cancellations *sync.Map // map[string]context.CancelFunc

	// idleWorkers is an unordered list of child worker IDs that
	// do not have work.
	idleWorkers []string

	// systemIdle is set if an attempt to get work returned
	// nothing.  In this case, there will not be another attempt
	// to get work for PollDuration time.
	systemIdle bool
}

var (
	// expirationWarning is a duration such that, if less than
	// this time is remaining to execute a work unit before it
	// expires, the worker will be signaled.
	expirationWarning = 30 * time.Second

	// expirationAlarm is a duration such that, if less than this
	// time is remaining to execute a work unit before it expires,
	// it will be failed instead.
	expirationAlarm = 15 * time.Second
)

// setDefaults sets default values for any Worker fields that are
// uninitialized.
func (w *Worker) setDefaults() {
	if w.WorkerID == "" {
		// May as well use a UUID here, "it's what we've always done"
		w.WorkerID = uuid.NewV4().String()
	}

	if w.Concurrency == 0 {
		w.Concurrency = runtime.NumCPU()
	}

	if w.PollInterval == time.Duration(0) {
		w.PollInterval = time.Duration(1) * time.Second
	}

	if w.HeartbeatInterval == time.Duration(0) {
		w.HeartbeatInterval = time.Duration(15) * time.Second
	}

	if w.MaxAttempts == 0 {
		w.MaxAttempts = 100
	}

	if w.Clock == nil {
		w.Clock = clock.New()
	}
}

// bootstrap creates the minimum required object set for the worker
// object.
func (w *Worker) bootstrap() error {
	w.childWorkers = make(map[string]coordinate.Worker)
	w.cancellations = new(sync.Map)

	// Get the parent worker
	var err error
	w.parentWorker, err = w.Namespace.Worker(w.WorkerID)
	return err
}

// Run runs work units from Coordinate forever, or until the provided
// context is cancelled.  If it returns, either there was a startup
// error connecting to Coordinate, in which case the corresponding
// error is returned, or execution was cancelled, returning nil.  If
// there is an error while trying to get attempts it is ignored.
func (w *Worker) Run(ctx context.Context) error {
	w.setDefaults()
	if err := w.bootstrap(); err != nil {
		return err
	}

	// This channel is signaled in doWork() after
	// RequestAttempts() returns, with a true value if at least
	// one attempt comes back.  If it does signal true, it
	// triggers another worker if possible.
	gotWork := make(chan bool)

	// This channel is signaled at the end of doWork() with the
	// worker ID.  If the most recent signal in gotWork was true,
	// it triggers another worker if possible.
	finished := make(chan string)

	// This channel, if non-nil, is signaled every second, and
	// triggers a worker.  It has a channel only when gotWork
	// transmits a false value.
	var tick <-chan time.Time
	var ticker *clock.Ticker

	// This channel is signaled to run the heartbeat task.
	var heartbeat <-chan time.Time
	heartbeater := w.Clock.Ticker(w.HeartbeatInterval)

	// We need to (asynchronously) kick off the world by telling
	// ourselves that it's okay to get more work units.
	go func() { gotWork <- true }()

	// TODO(dmaze): check for and signal stale workers
	//
	// The Python parent does both of these together every
	// heartbeat_interval (15 seconds).
	//
	// Also, the children in Python land register on creation and
	// unregister on destruction; but they only run a single work
	// unit or for 10 seconds.  Note that "expired" doesn't carry
	// a whole lot of meaning for workers, yet.

	for {
		select {
		case <-ctx.Done():
			// Shutting down
			// TODO(dmaze): don't return *immediately*; wait
			// for all of the workers to clean up
			if ticker != nil {
				ticker.Stop()
			}
			heartbeater.Stop()
			return nil

		case notIdle := <-gotWork:
			// Some worker came back with its
			// RequestAttempts() result.  If notIdle is
			// false, we expect the finished channel to
			// immediately be signaled as well.

			// If the "idle" bit changed, set/cancel the timer
			if w.systemIdle && notIdle {
				ticker.Stop()
				ticker = nil
				tick = nil
			}
			if (!w.systemIdle) && (!notIdle) {
				ticker = w.Clock.Ticker(w.PollInterval)
				tick = ticker.C
			}
			w.systemIdle = !notIdle
			w.maybeDoWork(ctx, gotWork, finished, false)

		case child := <-finished:
			// Some worker finished.
			w.returnIdleChild(child)
			w.maybeDoWork(ctx, gotWork, finished, false)

		case <-tick:
			// The system is idle, and the clock tick
			// fired.  Forcibly start an idle child at
			// this point.  This will eventually cause
			// gotWork to trigger, which will update the
			// systemIdle flag and might trigger even more
			// work.
			w.maybeDoWork(ctx, gotWork, finished, true)

		case <-heartbeat:
			w.heartbeat()
			w.findStaleUnits()
		}
	}
}

// getIdleChild returns the worker ID of a child worker that is not
// currently doing anything, or an empty string.  If the idle workers
// list is empty but the child workers list could have another worker,
// creates a new child.  Removes the returned worker from the idle
// workers list.
//
// Returns an empty string if there is no idle child and one cannot be
// created, including an error returned by the Coordinate backend.
func (w *Worker) getIdleChild() string {
	// Something in the idle workers list?  Just pick one
	if len(w.idleWorkers) > 0 {
		child := w.idleWorkers[0]
		w.idleWorkers = w.idleWorkers[1:]
		return child
	}

	// Can we support another worker?  Create one
	if len(w.childWorkers) < w.Concurrency {
		id := uuid.NewV4().String()
		child, err := w.Namespace.Worker(id)
		if err == nil {
			err = child.SetParent(w.parentWorker)
		}
		if err == nil {
			w.childWorkers[id] = child
			return id
		}
		if w.ErrorHandler != nil {
			w.ErrorHandler(err)
		}
		return ""
	}

	// Otherwise we're busy
	return ""
}

// returnIdleChild puts a child worker back into the idle children
// list, or if the system is idle, shuts down the worker.
func (w *Worker) returnIdleChild(id string) {
	child := w.childWorkers[id]
	if w.systemIdle {
		delete(w.childWorkers, id)
		w.cancellations.Delete(id)
		err := child.Deactivate()
		if err != nil && w.ErrorHandler != nil {
			w.ErrorHandler(err)
		}
	} else {
		w.idleWorkers = append(w.idleWorkers, id)
	}
}

// maybeDoWork spawns a new goroutine to do work, if there is an idle
// child worker.  If the system is idle, a new goroutine is never
// generated unless evenIfIdle is true.
func (w *Worker) maybeDoWork(ctx context.Context, gotWork chan<- bool, finished chan<- string, evenIfIdle bool) {
	if w.systemIdle && !evenIfIdle {
		return
	}
	child := w.getIdleChild()
	if child == "" {
		return
	}
	go w.doWork(ctx, child, w.childWorkers[child], gotWork, finished)
}

// doWork gets attempts and runs them.  It assumes it is running in its
// own goroutine.  It signals gotWork when the call to RequestAttempts
// returns, and signals finished immediately before returning.
func (w *Worker) doWork(ctx context.Context, id string, worker coordinate.Worker, gotWork chan<- bool, finished chan<- string) {
	// When we finish, signal the finished channel with our own ID
	defer func() {
		finished <- id
	}()

	attempts, err := worker.RequestAttempts(coordinate.AttemptRequest{
		Runtimes:          w.runtimes(),
		NumberOfWorkUnits: w.MaxAttempts,
	})
	if err != nil {
		// Handle the error if we can, but otherwise act just like
		// we got no attempts back
		if w.ErrorHandler != nil {
			w.ErrorHandler(err)
		}
		gotWork <- false
		return
	}
	if len(attempts) == 0 {
		// Nothing to do
		gotWork <- false
		return
	}
	// Otherwise we have actual work (and at least one attempt).
	gotWork <- true

	// See if we can find a task for the work spec
	spec := attempts[0].WorkUnit().WorkSpec()
	task := spec.Name()
	data, err := spec.Data()
	if err == nil {
		aTask, present := data["task"]
		if present {
			bTask, ok := aTask.(string)
			if ok {
				task = bTask
			}
		}
	}

	// Try to find the task function
	var taskFn func(context.Context, []coordinate.Attempt)
	if err == nil {
		taskFn = w.Tasks[task]
		if taskFn == nil {
			err = fmt.Errorf("No such task function %q", task)
		}
	}

	if err == nil {
		taskCtx, cancellation := context.WithCancel(ctx)
		w.cancellations.Store(id, cancellation)
		taskFn(taskCtx, attempts)
		// It appears to be recommended to call this; calling
		// it multiple times is documented to have no effect
		cancellation()
	} else {
		failure := map[string]interface{}{
			"traceback": err.Error(),
		}
		// Try to fail all the attempts, ignoring errors
		for _, attempt := range attempts {
			_ = attempt.Fail(failure)
		}
	}
}

// runtimes returns the configured runtimes for this worker. If no
// runtimes are configured, then a default value, ["go"], is returned.
func (w *Worker) runtimes() []string {
	if w.Runtimes == nil {
		return []string{"go"}
	}
	return w.Runtimes
}

// heartbeat reports the current status of the parent worker.
func (w *Worker) heartbeat() {
	data := map[string]interface{}{
		"cpus":       runtime.NumCPU(),
		"go":         runtime.Version(),
		"goroutines": runtime.NumGoroutine(),
		"pid":        os.Getpid(),
	}
	hostname, err := os.Hostname()
	if err == nil {
		data["hostname"] = hostname
	}
	interfaces, err := net.Interfaces()
	if err == nil {
		var ipaddrs []string
		for _, interf := range interfaces {
			addrs, err := interf.Addrs()
			if err == nil {
				for _, addr := range addrs {
					ipaddrs = append(ipaddrs, addr.String())
				}
			}
		}
		if len(ipaddrs) > 0 {
			data["ipaddrs"] = ipaddrs
		}
	}
	// The Python worker publishes quite a bit of extra
	// information here: "version" is the version of the
	// coordinate package; "working_set" is the complete pip
	// environment; "memory" is a dictionary with system memory
	// statistics.
	//
	// I think this is rarely examined.  Some of it is hard to get
	// portably.  Even the information that's retrieved above
	// could be uninformative if a worker is run in a Docker
	// container (e.g. the pid will always be 1 and the IP address
	// an uninformative host-local address).

	now := w.Clock.Now()
	then := now.Add(time.Duration(15) * time.Minute)
	err = w.parentWorker.Update(data, now, then, "RUN")
	if err != nil && w.ErrorHandler != nil {
		w.ErrorHandler(err)
	}
}

// findStaleUnits looks for work units assigned to child workers of w
// that are about to run out of time, and cancels them.
func (w *Worker) findStaleUnits() {
	// Scan our childrens' attempts to see if any are expiring
	attempts, err := w.parentWorker.ChildAttempts()
	if err != nil {
		return
	}
	now := w.Clock.Now()
	childrenToCancel := make(map[string]struct{})
	for _, attempt := range attempts {
		exp, err := attempt.ExpirationTime()
		if err == nil {
			remaining := exp.Sub(now)
			if remaining < expirationWarning {
				// TODO(dmaze): if the attempt just
				// has a really short lease time, we
				// should let at least half its
				// possible time elapse
				child := attempt.Worker().Name()
				childrenToCancel[child] = struct{}{}
			}
			if remaining < expirationAlarm {
				// Proactively fail the attempt
				err = attempt.Fail(map[string]interface{}{
					"traceback": "timed out",
				})
			}
		}
	}

	// If anything is expiring (we have a list of child worker IDs,
	// not specific attempts) call their cancellation functions
	for child := range childrenToCancel {
		cancellation, ok := w.cancellations.Load(child)
		// *should* always be there but doesn't hurt to check
		if ok {
			cancellation.(context.CancelFunc)()
		}
	}
}
